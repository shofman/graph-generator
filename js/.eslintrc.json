{
    "env": {
        "browser": true,
        "es6": true
    },
    "extends": ["prettier", "eslint:recommended"],
    "parserOptions": {
        "sourceType": "module"
    },
    "plugins": ["prettier", "jest"],
    "rules": {
        "prettier/prettier": "error",
        "indent": [
            "error",
            2
        ],
        "no-console": "off",
        "linebreak-style": [
            "error",
            "unix"
        ],
        "quotes": [
            "error",
            "single"
        ],
        "semi": [
            "error",
            "never"
        ]
    },
    "env": {
        "jest/globals": true,
        "browser": true,
        "node": true
    }
}
