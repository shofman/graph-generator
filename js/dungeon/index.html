<html lang="en">
<head>
    <script type="text/javascript" src="./ui/vis.js"></script>
    <link href="./ui/vis.css" rel="stylesheet" type="text/css" />

    <style type="text/css">
        #generation, #mynetwork {
            float: left;
        }
        #generation2 {
            display: block;
        }
        #dungeonVisual, #dungeonRooms {
            border: 1px solid black;
            float: left;
        }

        #dungeonVisual {
            max-width: 1000px;
            min-width: 400px;
        }
        .space-left {
            margin-left: 18px;
        }
        .dungeon-grouping {
            display: flex;
            flex-direction: column;
        }
        .dungeon-header {
            flex-direction: row;
        }
        table {
            table-layout: fixed;
            width: 100%;
        }
    
        table td {
            width: 50px;   
            text-align: center; 
            vertical-align: middle;
            border: 1px solid lightgray;
            max-width: 80px;
            word-wrap: break-word;
        }

        td.filled {
            background: lightgray;
            border: 4px solid black;
            padding: 10px;
        }

        td.empty {
            display: none;
        }

        td.lower-parent {
            border-bottom: 0px solid black;
        }

        td.upper-child {
            border-top: 0px solid black;
        }

        td.left-child {
            border-left: 0px solid black;
        }

        td.right-parent {
            border-right: 0px solid black;
        }

        td.right-child {
            border-right: 0px solid black;
        }

        td.left-parent {
            border-left: 0px solid black;
        }

        td.lower-child {
            border-bottom: 0px solid black;
        }

        td.upper-parent {
            border-top: 0px solid black;
        }
        .dungeon-room-visual {
            display: flex;
        }
        #roomdebugger {
            display: flex;
            flex-direction: column;
            float: left;
        }
    </style>
</head>
<body>
    
<script type="module" src="./main.js"></script>

<button id="rewind" onclick="rewind()">-</button>
<input id="generation" type="range" min="0" range="1" onchange="drawDungeonTree()"/>
<button id="advance" onclick="advance()">+</button>
<input id="keys" type="checkbox" onchange="drawDungeonTree()"/>
<button id="verify" onclick="verify()">Verify</button>
<button id="evaluate" onclick="evaluateDungeon()">Evaluate</button>
<span id="seed"></span>
<button id="generation2" onclick="drawDungeonTree()" type="button">Refresh</button>

<div id="roomdebugger">
    <div id="mynetwork"></div>
    <canvas id="dungeonRooms" width="800" height="400"></canvas>
</div>

<div class="dungeon-grouping">
    <div class="dungeon-header">
        <label class="space-left" for="dungeonView">View rooms/groups</label>
        <input id="dungeonView" type="checkbox" onchange="switchViews()">
        <button id="rewindDraw" onclick="rewindDraw()">-</button>
        <input id="drawDungeonSlider" type="range" min="0" range="1" value="0" onchange="stepThroughDungeon()" />
        <button id="advanceDraw" onclick="advanceDraw()">+</button>
    </div>
    <div class="dungeon-room-visual">
        <table class="space-left" id="dungeonVisual" width="650" height="800"></table>
    </div>
</div>

</body>
</html>
