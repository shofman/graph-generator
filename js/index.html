<html>
<head>
    <script type="text/javascript" src="./vis.js"></script>
    <link href="./vis.css" rel="stylesheet" type="text/css" />

    <style type="text/css">
        #generation, #mynetwork {
            display: inline;
        }
        #generation2 {
            display: block;
        }
    </style>
</head>
<body>
<script type="text/javascript" src="./generateDungeon.js"></script>
<button id="rewind" onclick="rewind()">-</button>
<input id="generation" type="range" min="1" max="15" range="1" value="15" onchange="drawDungeon()"/>
<button id="advance" onclick="advance()">+</button>
<input id="keys" type="checkbox" onchange="drawDungeon()"/>
<button id="verify" onclick="verify()">Verify</button>
<span id="seed"></span>
<button id="generation2" onclick="drawDungeon()" type="button">Refresh</button>

<div id="mynetwork"></div>

<image id="imagegraph" src="./watertemple.png" width="800" height="800" />
<!-- <image id="imagegraph" src="./zelda+dungeon2.jpg_large" width="800" height="800" /> -->

<script type="text/javascript">
    drawDungeon()
</script>
</body>
</html>
