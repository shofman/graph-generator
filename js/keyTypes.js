const KEY_TYPES = {
  BOSS: 'boss',
  NORMAL_KEY: 'key',
  KEY_ITEM: 'key_item',
  MULTI_KEY: 'multiKey',
  MULTI_LOCK: 'multiLock',
  SINGLE_LOCK_KEY: 'singleKeyLock',
  UNKNOWN: 'unknown',
}
