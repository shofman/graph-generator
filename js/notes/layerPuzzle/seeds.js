// Second grid seeds
export const broken = 0.13211651171095662
export const weird = 0.8814176675958372
export const huh = 0.9414135727191293
export const solved = 0.9311797150098673
export const needToFixBacktracking = 0.7037516700656562

// Third grid seeds
export const infinite = 0.8270123205982178
export const infinite2 = 0.5020871473412547
export const fucked = 0.45286728697492507

export const tiedLargest = 0.2996170901896136
export const breakThisDown = 0.949010289749602
export const howToHandle = 0.7210721903502999
export const badTargetPlacement = 0.27211505962386684
export const chunky = 0.6757205201354943
export const multipleTargetsWithEvenLengths = 0.33285831629572726
export const noTargets = 0.4598295234421774
export const reallyLongTail = 0.8742797946667888
export const handle = 0.16586225158162837
export const brokenAgain = 0.8807449728181744
export const reallyDense = 0.006649334400887774
export const goodSeed = 0.16134528824178074
export const fuckingUseless = 0.6013755322878005
export const superBoring = 0.41802797071231046
export const generateLeftTargetNodeFromNothing = 0.1228821185565192

// Seeds with backtracking implemented that should work
export const tooSimpleSolution = 0.589150276446149
export const solutionRightBeside = 0.07279613399399987
