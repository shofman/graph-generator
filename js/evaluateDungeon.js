import { getDistanceBetweenLeafs, getLockedDistanceBetweenLeafs } from './calculateLeafDistances.js'
import { drawDungeonTree } from './ui.js'
import { verifyDungeons } from './verifyDungeon.js'
import { KEY_TYPES } from './keyTypes.js'

export const isMeaningfulBranch = node => {
  const nodeHasBranchingBinaryChildren = node => {
    return node.children.length === 2 && node.children.every(child => child.children.length >= 2)
  }
  return node.hasChildren() && (node.children.length >= 3 || nodeHasBranchingBinaryChildren(node))
}

// TODO - try using dfs and comparing the results (because players will likely do both - bfs is unlikely to be the sole/popular strategy)

const calculateVisitedPathCost = visitedPath => {
  window.visitedPath = visitedPath
  let index1 = 0
  let index2 = 1

  let distance = 0

  while (index2 < visitedPath.length) {
    const firstNode = visitedPath[index1++]
    const secondNode = visitedPath[index2++]

    distance += getDistanceBetweenLeafs(firstNode, secondNode).distance
  }

  return distance
}

window.calcPath = calculateVisitedPathCost

export const calculateDungeonScore = dungeon => {
  let dungeonInfo = {
    aName: dungeon.dungeon.seedName || dungeon.seedName,
    keys: {
      bossKey: {
        distanceToStart: 0,
        distanceToGate: 0,
        lockedDistance: 0,
        lockedDistanceToStart: 0,
      },
      keyItem: {
        distanceToStart: 0,
        avgDistanceToGate: 0,
        avgLockedDistance: 0,
      },
      normalKey: {
        avgDistance: 0,
        lockedDistance: 0,
      },
      specialKey: {
        avgDistance: 0,
        avgLockedDistance: 0,
      },
      multiKey: {
        avgDistance: 0,
        avgLockedDistance: 0,
      },
      multiLock: {
        avgDistance: 0,
        avgLockedDistance: 0,
      },
    },
    numberOfMeaningfulBranches: 0,
    treeStructure: {},
    maxHeight: 0,
    visitedPath: dungeon.visitedPath,
    criticalPathDistance: calculateVisitedPathCost(dungeon.visitedPath),
    numberOfNodes: dungeon.visitedPath.length,
  }

  dungeon.visitedPath.forEach(node => {
    if (isMeaningfulBranch(node)) {
      dungeonInfo.numberOfMeaningfulBranches++
    }

    const nodeHeight = node.calculateHeight()
    if (nodeHeight > dungeonInfo.maxHeight) {
      dungeonInfo.maxHeight = nodeHeight
    }
    if (dungeonInfo.treeStructure[nodeHeight]) {
      dungeonInfo.treeStructure[nodeHeight] += 1
    } else {
      dungeonInfo.treeStructure[nodeHeight] = 1
    }
  })

  Object.keys(dungeon.keysGroupedByType).forEach(classifiedNodeKeys => {
    const currentNodeGroup = dungeon.keysGroupedByType[classifiedNodeKeys]
    if (currentNodeGroup.length) {
      switch (classifiedNodeKeys) {
        case KEY_TYPES.BOSS: {
          currentNodeGroup.forEach(key => {
            const bossGate = key.locks[0]
            dungeonInfo.keys.bossKey.distanceToGate = getDistanceBetweenLeafs(
              bossGate,
              key
            ).distance
            dungeonInfo.keys.bossKey.lockedDistance = getLockedDistanceBetweenLeafs(key, bossGate)
            dungeonInfo.keys.bossKey.distanceToStart = key.calculateHeight()
            dungeonInfo.keys.bossKey.lockedDistanceToStart = getLockedDistanceBetweenLeafs(
              key,
              dungeon.visitedPath[0]
            )
          })
          break
        }

        case KEY_TYPES.KEY_ITEM: {
          let totalKeyItemDistanceFromStart = 0
          currentNodeGroup.forEach(key => {
            totalKeyItemDistanceFromStart += key.calculateHeight()

            const keyItemLocks = key.locks
            keyItemLocks.forEach(keyItemLock => {
              dungeonInfo.keys.keyItem.avgDistanceToGate += getDistanceBetweenLeafs(
                keyItemLock,
                key
              ).distance
              dungeonInfo.keys.keyItem.avgLockedDistance += getLockedDistanceBetweenLeafs(
                keyItemLock,
                key
              )
            })

            dungeonInfo.keys.keyItem.avgDistanceToGate =
              dungeonInfo.keys.keyItem.avgDistanceToGate / keyItemLocks.length
            dungeonInfo.keys.keyItem.avgLockedDistance =
              dungeonInfo.keys.keyItem.avgLockedDistance / keyItemLocks.length
          })

          dungeonInfo.keys.keyItem.distanceToStart =
            totalKeyItemDistanceFromStart / currentNodeGroup.length
          break
        }

        case KEY_TYPES.MULTI_KEY: {
          currentNodeGroup.forEach(key => {
            const locks = key.locks
            locks.forEach(lock => {
              dungeonInfo.keys.multiKey.avgDistance += getDistanceBetweenLeafs(lock, key).distance
              dungeonInfo.keys.multiKey.avgLockedDistance += getLockedDistanceBetweenLeafs(
                lock,
                key
              )
            })
          })

          dungeonInfo.keys.multiKey.avgDistance =
            dungeonInfo.keys.multiKey.avgDistance / currentNodeGroup.length
          dungeonInfo.keys.multiKey.avgLockedDistance =
            dungeonInfo.keys.multiKey.avgLockedDistance / currentNodeGroup.length
          break
        }

        case KEY_TYPES.MULTI_LOCK: {
          currentNodeGroup.forEach(key => {
            const locks = key.locks
            locks.forEach(lock => {
              dungeonInfo.keys.multiLock.avgDistance += getDistanceBetweenLeafs(lock, key).distance
              dungeonInfo.keys.multiLock.avgLockedDistance += getLockedDistanceBetweenLeafs(
                lock,
                key
              )
            })

            dungeonInfo.keys.multiLock.avgDistance =
              dungeonInfo.keys.multiLock.avgDistance / locks.length
            dungeonInfo.keys.multiLock.avgLockedDistance =
              dungeonInfo.keys.multiLock.avgLockedDistance / locks.length
          })

          break
        }

        case KEY_TYPES.NORMAL_KEY: {
          currentNodeGroup.forEach(key => {
            const locks = key.locks
            locks.forEach(lock => {
              dungeonInfo.keys.normalKey.avgDistance += getDistanceBetweenLeafs(lock, key).distance
              dungeonInfo.keys.normalKey.lockedDistance += getLockedDistanceBetweenLeafs(lock, key)
            })
          })
          dungeonInfo.keys.normalKey.avgDistance =
            dungeonInfo.keys.normalKey.avgDistance / currentNodeGroup.length
          dungeonInfo.keys.normalKey.lockedDistance =
            dungeonInfo.keys.normalKey.lockedDistance / currentNodeGroup.length
          break
        }

        case KEY_TYPES.COMBAT_KEY:
        case KEY_TYPES.SINGLE_LOCK_KEY:
          currentNodeGroup.forEach(key => {
            const locks = key.locks
            locks.forEach(lock => {
              dungeonInfo.keys.specialKey.avgDistance += getDistanceBetweenLeafs(lock, key).distance
              dungeonInfo.keys.specialKey.avgLockedDistance += getLockedDistanceBetweenLeafs(
                lock,
                key
              )
            })

            dungeonInfo.keys.specialKey.avgDistance =
              dungeonInfo.keys.specialKey.avgDistance / locks.length
            dungeonInfo.keys.specialKey.avgLockedDistance =
              dungeonInfo.keys.specialKey.avgLockedDistance / locks.length
          })
          break

        case KEY_TYPES.SINGLE_ROOM_PUZZLE:
        case KEY_TYPES.EXTERNAL_KEY:
          break

        default: {
          break
        }
      }
    }
  })

  let avgTreeWidth = 0
  let maxTreeWidth = 0

  Object.values(dungeonInfo.treeStructure).forEach(levelResult => {
    avgTreeWidth += levelResult
    if (levelResult > maxTreeWidth) {
      maxTreeWidth = levelResult
    }
  })

  avgTreeWidth /= Object.values(dungeonInfo.treeStructure).length

  return Object.assign(dungeonInfo, { maxTreeWidth, avgTreeWidth })
}

export const evaluateDungeon = () => {
  const verifiedDungeons = verifyDungeons(drawDungeonTree(true))
  if (verifiedDungeons) {
    const evaluations = verifiedDungeons
      .filter(dungeon => !!dungeon.visitedPath.length)
      .map(calculateDungeonScore)

    const calcScore = (aScore, bScore) => {
      if (aScore > bScore) {
        return 1
      } else if (aScore < bScore) {
        return -1
      }
      return 0
    }

    const sortByNormalizedCriticalPath = (a, b) => {
      const aScore = a.criticalPathDistance / a.numberOfNodes
      const bScore = b.criticalPathDistance / b.numberOfNodes
      return calcScore(aScore, bScore)
    }

    // This one does not seem to be as important - obviously, we want increase this value, but not as important as critical path
    const sortByBossKey = (a, b) => {
      const aKey = a.keys.bossKey
      const aScore =
        (aKey.distanceToStart + aKey.distanceToGate) / a.maxHeight +
        (aKey.lockedDistance + aKey.lockedDistanceToStart) / a.numberOfNodes

      const bKey = b.keys.bossKey
      const bScore =
        (bKey.distanceToStart + bKey.distanceToGate) / b.maxHeight +
        (bKey.lockedDistance + bKey.lockedDistanceToStart) / b.numberOfNodes
      return calcScore(aScore, bScore)
    }

    const sortByKeyItem = (a, b) => {
      const aKey = a.keys.keyItem
      const aScore =
        (aKey.distanceToStart + aKey.avgDistanceToGate) / a.maxHeight +
        aKey.avgLockedDistance / a.numberOfNodes

      const bKey = b.keys.keyItem
      const bScore =
        (bKey.distanceToStart + bKey.avgDistanceToGate) / b.maxHeight +
        bKey.avgLockedDistance / b.numberOfNodes

      return calcScore(aScore, bScore)
    }

    const sortBySumOfKeys = (a, b) => {
      const aKey = a.keys
      const aDistanceAvg =
        (aKey.normalKey.avgDistance +
          aKey.specialKey.avgDistance +
          aKey.multiKey.avgDistance +
          aKey.multiLock.avgDistance) /
        a.numberOfNodes
      const aLockedDistance =
        (aKey.normalKey.lockedDistance +
          aKey.specialKey.lockedDistance +
          aKey.multiLock.lockedDistance +
          aKey.multiKey.lockedDistance) /
        a.numberOfNodes

      const bKey = b.keys
      const bDistanceAvg =
        (bKey.normalKey.avgDistance +
          bKey.specialKey.avgDistance +
          bKey.multiKey.avgDistance +
          bKey.multiLock.avgDistance) /
        b.numberOfNodes
      const bLockedDistance =
        (bKey.normalKey.lockedDistance +
          bKey.specialKey.lockedDistance +
          bKey.multiLock.lockedDistance +
          bKey.multiKey.lockedDistance) /
        b.numberOfNodes

      const aScore = aDistanceAvg + aLockedDistance
      const bScore = bDistanceAvg + bLockedDistance

      return calcScore(aScore, bScore)
    }

    const noLog = () => {}
    evaluations.sort(sortByBossKey).forEach(dungeon => noLog(dungeon))
    evaluations.sort(sortByKeyItem).forEach(dungeon => noLog(dungeon))
    evaluations.sort(sortBySumOfKeys).forEach(dungeon => noLog(dungeon))
    evaluations
      .sort(sortByNormalizedCriticalPath)
      .forEach(dungeon =>
        console.log(dungeon, dungeon.criticalPathDistance / dungeon.numberOfNodes)
      )
  }

  // Distance needs to take into account the verticality / # of locked doors in the way
  // E.g. if two nodes share a parent, that should be scored differently than if two nodes are separated by a locked door
  // With distance as the only measurement, these would both be a 2

  // Handle the key scoring better - keys are universal, so distance to unlock may not be indicative of true difficulty

  // Should I include the aggregate distances? E.g. if a boss key and gate are separated by a lot of locks, should I add these to the scores

  // Create two 'wandering' BFS. One right to left, other left to right. Avg the score (or just randomize the ordering of the nodes)

  // so far Critical path btwn boss key and gate  is always composed of several gates (4-6)

  // Can evaluate how we are doing at each step of the creation phase to determine how aggressive the generation needs to be

  // Handle first key assumption in lockedDistance func
  // Also compare the use of recursion or not

  // distance between boss key and where it is used - highest priority
  // distance between item and where it is used - higher priority than keys
  // distance between key and where it is used - lower priority than item
  // distance between boss key and entrance. Should be greater than 1
  // Total branching (how many children exist at a level). Water temple has a max of eight
  // Complete path length - between 17 -27 segments long

  // Add entry that looks at distance between types of keys (ensuring there's less clustering of puzzles/external locks in an area)

  // Ensure that the ratio between combat rooms and puzzle rooms are evenly distributed
}
