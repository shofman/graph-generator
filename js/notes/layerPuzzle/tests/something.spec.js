import { something } from '../seeds.js'
import { runTest } from './puzzleSolver.js'

jest.mock('../draw.js', () => ({
  debugDraw: jest.fn(),
}))
global.console = { log: jest.fn() }

describe('Handles slow solution', () => {
  it('matches snapshot for something', () => {
    const solvedResults = runTest(something)

    expect(solvedResults.grid).toMatchSnapshot()
    expect(solvedResults.results).toMatchSnapshot()
  })
})
