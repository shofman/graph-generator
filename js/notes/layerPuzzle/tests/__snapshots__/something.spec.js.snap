// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handles slow solution matches snapshot for something 1`] = `undefined`;

exports[`Handles slow solution matches snapshot for something 2`] = `
Object {
  "10,2": Object {
    "blocksHandledByLine": Array [
      "10,3",
      "10,4",
    ],
    "blocksHandledByOthers": Array [],
    "direction": "vertical",
    "key": "10,2",
    "lineDelta": Object {
      "x": 0,
      "y": 1,
    },
    "solutionOrderNumber": 7,
    "value": 2,
  },
  "10,5": Object {
    "blocksHandledByLine": Array [
      "10,6",
    ],
    "blocksHandledByOthers": Array [],
    "direction": "vertical",
    "key": "10,5",
    "lineDelta": Object {
      "x": 0,
      "y": 1,
    },
    "solutionOrderNumber": 10,
    "value": 1,
  },
  "11,2": Object {
    "blocksHandledByLine": Array [
      "5,2",
      "6,2",
      "8,2",
      "3,2",
    ],
    "blocksHandledByOthers": Array [
      "10,2",
      "9,2",
      "7,2",
      "4,2",
    ],
    "direction": "horizontal",
    "key": "11,2",
    "lineDelta": Object {
      "x": -1,
      "y": 0,
    },
    "solutionOrderNumber": 1,
    "value": 4,
  },
  "11,3": Object {
    "blocksHandledByLine": Array [
      "9,3",
    ],
    "blocksHandledByOthers": Array [
      "10,3",
    ],
    "direction": "horizontal",
    "key": "11,3",
    "lineDelta": Object {
      "x": -1,
      "y": 0,
    },
    "solutionOrderNumber": 5,
    "value": 1,
  },
  "11,4": Object {
    "blocksHandledByLine": Array [
      "9,4",
    ],
    "blocksHandledByOthers": Array [
      "10,4",
    ],
    "direction": "horizontal",
    "key": "11,4",
    "lineDelta": Object {
      "x": -1,
      "y": 0,
    },
    "solutionOrderNumber": 6,
    "value": 1,
  },
  "11,6": Object {
    "blocksHandledByLine": Array [
      "9,6",
    ],
    "blocksHandledByOthers": Array [
      "10,6",
    ],
    "direction": "horizontal",
    "key": "11,6",
    "lineDelta": Object {
      "x": -1,
      "y": 0,
    },
    "solutionOrderNumber": 9,
    "value": 1,
  },
  "4,0": Object {
    "blocksHandledByLine": Array [
      "4,1",
      "4,2",
    ],
    "blocksHandledByOthers": Array [],
    "direction": "vertical",
    "key": "4,0",
    "lineDelta": Object {
      "x": 0,
      "y": 1,
    },
    "solutionOrderNumber": 2,
    "value": 2,
  },
  "7,4": Object {
    "blocksHandledByLine": Array [
      "7,3",
      "7,2",
    ],
    "blocksHandledByOthers": Array [],
    "direction": "vertical",
    "key": "7,4",
    "lineDelta": Object {
      "x": 0,
      "y": -1,
    },
    "solutionOrderNumber": 3,
    "value": 2,
  },
  "8,5": Object {
    "blocksHandledByLine": Array [
      "9,5",
    ],
    "blocksHandledByOthers": Array [],
    "direction": "horizontal",
    "key": "8,5",
    "lineDelta": Object {
      "x": 1,
      "y": 0,
    },
    "solutionOrderNumber": 8,
    "value": 1,
  },
  "9,7": Object {
    "blocksHandledByLine": Array [
      "9,2",
    ],
    "blocksHandledByOthers": Array [
      "9,6",
      "9,5",
      "9,4",
      "9,3",
    ],
    "direction": "vertical",
    "key": "9,7",
    "lineDelta": Object {
      "x": 0,
      "y": -1,
    },
    "solutionOrderNumber": 4,
    "value": 1,
  },
}
`;
